hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpcengine/nio/NioAsyncSocket.java

```diff
@@ -424,28 +470,48 @@ private void handleRead() throws IOException {
             compactOrClear(rcvBuffer);
         }
 
+        // todo: temp notes.
+        // In netty there is logic in the NioSocketChannel.doWrite
+        // 1) all the collected messages are IOBUffers, do a writev.
+        // 2) if there is only 1 message and it an IOBuffer, then do a write
+        // 3) if at least one of the messages is a non IOBuffer, then fallback
+        // to a normal write.
         private void handleWrite() throws IOException {
-            // typically this method is called with the flushThread being set.
-            // but in case of cancellation of the key, this method is also
-            // called without the flushThread being set.
-            // So we can't do an assert flushThread!=null.
-
             metrics.incWriteEvents();
 
-            ioVector.populate(writeQueue);
+            // todo: Netty has a write spin option we need to investigate.
 
-            ByteBuffer[] srcs = ioVector.array();
-            int length = ioVector.length();
-            long written = length == 1
-                    ? socketChannel.write(srcs[0])
-                    : socketChannel.write(srcs, 0, length);
+            long bytesWritten;
+            boolean clean;
+            if (writer == null) {
+                // the writeQueue is guaranteed to have only IOBuffers
+                // if the writer isn't set.
+                ioVector.populate(writeQueue);
 
-            ioVector.compact(written);
+                if (writeQueue.isEmpty()) {
+                    ioVectorWriteAllowed = true;
+                }
 
-            metrics.incBytesWritten(written);
-            //System.out.println(NioAsyncSocket.this + " bytes written:" + written);
+                int ioVectorLength = ioVector.length();
+                ByteBuffer[] srcs = ioVector.array();
+                bytesWritten = ioVectorLength == 1
+                        ? socketChannel.write(srcs[0])
+                        : socketChannel.write(srcs, 0, ioVectorLength);
+                ioVector.compact(bytesWritten);
+                clean = ioVector.isEmpty();
+            } else {
+                boolean writerClean = writer.onWrite(sndBuffer);
+                sndBuffer.flip();
+                bytesWritten = socketChannel.write(sndBuffer);
+                boolean sndBufferClean = !sndBuffer.hasRemaining();
+                clean = writerClean && sndBufferClean;
+                compactOrClear(sndBuffer);
+            }
 
-            if (ioVector.isEmpty()) {
+            metrics.incBytesWritten(bytesWritten);
+            //System.out.println(socket + " bytes written:" + bytesWritten);
```

Comment:
Should these be included?

Reply:
Yes. Extremely helpful when there are problems which need to be debugged.